development/workflow.rst

```diff
@@ -0,0 +1,305 @@
+
+Associating commits to a single logical changes debugging and code audits easer:
+
+- Git bisect is more effective for zeroing in on the change that introduced a regression.
+- Git blame is more helpful for explaining why a change was made.
+- Better commit organization guides reviewers through your pull request, making for more effective code reviews.
+- A bad commit can more easily be reverted later with fewer side-effects.
+
+Some edits serve only to fix white space or code style issues in existing code.
+Those whitespace and style fixes should be made in separate commits from new development.
+Usually it makes sense to fix whitespace and style issues in code *before* embarking on new development.
+
+Rebase commits from code reviews rather than having 'review fixes' commits
+--------------------------------------------------------------------------
+
+Code review will result in additional commits that address code style, documentation and implementation issues.
+Authors should rebase (i.e., ``git rebase -i master``) their ticket branch to squash the post-review fixes to the pre-review commits.
+The end-goal is that a pull request, when merged, should have a commit history that tells a coherent development story and looks as if the code was written correctly the first time.
+
+There is *no need* to retain post-review commits in order to preserve code review discussions.
+So long as comments, even patch-level comments for out-of-date commits, are made in a pull request, GitHub will preserve that content.  
+
+.. _git-commit-message-best-practices:
+
+Appendix: Commit Message Best Practices
+=======================================
+
+Generally you should write your commit messages in an editor, not at the prompt.
+Reserve the ``git commit -m "messsage"`` pattern for 'work in progress' commits that will be rebased before code review.
+
+There are strong conventions for Git commit messages that we follow.
+**Messages start with a single-line summary of 50 characters or less**.
+Consider 50 characters as a hard limit; your summary will by truncated in the  GitHub display otherwise.
+Write the message in the **imperative** tense, not the past tense.
+For example, "Add feature ..." or "Fix issue ..." rather than "Added feature..." or "Fixed feature...."
+Ensure the summary line contains the right keywords so that someone examining `a commit listing <https://github.com/lsst/afw/commits/master>`_ can understand what parts of the codebase are being changed.
+
+**The message body should be wrapped at 72 character line lengths**, and can contain a mix of lists and paragraphs that explain the code changes. 
+Use the body of a commit message to describe:
+
+- What the original issue was; the reader shouldn't have to look at JIRA to understand what prompted the code change.
+- What the changes actually are and why they were made.
+- What the limitations of the code are. This is especially useful for future debugging.
```

Comment:
If appropriate, what other approaches were considered for the problem and why they weren't used.